### ISSUE TITLE
test-coverage-constitutional-standard: counts 1 test file where 27 match its own isTestFile

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#93]
**Version:** 7.20.0
**Law:** `test-coverage-constitutional-standard` (`TestFilesExistenceChecker`)

## What happens

```
Low test coverage ratio: 1.6% (1 tests for 62 source files, minimum 50%)
62 source files lack corresponding test files
```

The repository has **27 files** that the checker's own `isTestFile` accepts — 9
Jest specs under `apps/web/src` and 18 Playwright specs under
`apps/web-e2e/src`, all named `*.spec.ts`. It counts **one**.

Source discovery works: it finds 62 testable sources, which matches this repo
(69 `.ts` files under `apps/`, less the ones `shouldHaveTest` skips). Only the
test side is wrong, so the ratio is computed from a numerator that is off by 26.

## Reproduction

Walking the tree with the checker's own predicate and its own skip list:

```js
// isTestFile, verbatim from test-files-existence-checker.js
fileName.includes('.test.') || fileName.includes('.spec.') || fileName.includes('__tests__')

// shouldSkipDirectory, verbatim
['node_modules','dist','build','.git','coverage','.nx','.angular','tmp','temp']
```

gives **27** files. The audit reports **1**.

## What I ruled out

- **Our ignore config.** `ignores.global` is nine build-artefact globs
  (`**/node_modules/**`, `**/dist/**`, `**/coverage/**`, `**/.nx/**`,
  `**/.angular/**`, `**/.ruleofcode-cache/**`, `**/tmp/**`, `**/*vite*/**`,
  `**/*chunk*/**`). None can match `apps/web/src/app/app.spec.ts`.
  `ignores.byRule` is empty and there is no `includes` whitelist.
- **The `includeTests` flag.** `findTestFiles` passes `true`, and
  `getIgnorePatterns` only appends `PATTERNS.SPEC_FILES` / `PATTERNS.TEST_FILES`
  / `**/*-e2e/**` when `includeTests` is falsy — so on this path the spec
  patterns are not in the ignore set.
- **The directory walk.** `shouldSkipDirectory` does not exclude `apps`,
  `web-e2e` or `src`.

I could not isolate the mechanism from reading the compiled output, so this
reports the asymmetry rather than a diagnosis. The remaining candidates I could
not inspect are the law-specific patterns `getIgnorePatterns` may add from
`lawId`, and anything upstream that scopes `projectRoot` differently for the two
scans.

## Why it matters here

This law is one of the reasons we hold a law at `warning` instead of `error`,
and our policy is that a law is `error` or it is declared not-applicable with a
written reason — `warning` is not somewhere a law should live. The finding as
reported is not payable: adding tests cannot move a number that is not counting
the tests already there.

It also reads as the most damning line on a public audit page — a site that
publishes its own debt showing "1.6% coverage" while running 27 spec files.

## Suggested direction

If the two scans resolve their ignore set differently, having them share one
call would make the asymmetry impossible. Failing that, listing the discovered
test files under `--verbose` would let an adopter see which 26 went missing
without decompiling the checker.

**Comment by ruleofcodedev** [comment]
**Fixed — in #97, awaiting CI and merge to `develop`. Ships in the next release.**

Your hypothesis was right and it was the first one you listed: the two scans resolved different ignore sets.

The recursive walk called the directory filter **without** `includeTests`, so it resolved the set for a scan that *excludes* tests — which adds `**/*-e2e/**`, `**/e2e/**`, `**/test/**`, `**/tests/**` and `**/cypress/**`. Whole directories were pruned before the file validator ever ran. You ruled out `includeTests` correctly for `findTestFiles` — it passes `true` there and that half was fine. The flag was missing one level up, in the walk.

Measured against the previous build, same fixtures:

| layout | before | after |
| --- | --- | --- |
| six typical spec locations | **1 of 6** | 6 of 6 |
| Nx: 9 unit + 18 Playwright | **9 of 27** | 27 of 27 |

Your "counts 1" reproduces exactly.

Your second request is in as well: the audit now names the test files it counted, with the line *"If a test you have is missing from this list, the ratio is wrong rather than your coverage."* You should not have had to re-implement the scan against `dist/` to see that, and nobody should have to again.

Worth saying: your report was usable precisely because you separated what you had verified from what you could not isolate. The three things you ruled out were all correctly ruled out.

**Comment by ruleofcodedev** [comment]
Merged to `develop` in #97. Ships in the next release.

**Comment by ruleofcodedev** [comment]
Thank you — four of the five landed here cleanly, confirmed by findings that
disappeared rather than by the notes: #94, #95, #96 and #98 are all good.

This one is not fixed on our side, and I can now show why more precisely than
the original report. **The checker is fixed. Something between it and the audit
is not.**

## The checker returns the right answer

Called directly, on RoC 7.21.0, in the repository the original report came from:

```js
const { TestFilesExistenceChecker: C } =
  require('ruleofcode/dist/utils/testing/test-files-existence-checker.js');

C.findTestFiles(process.cwd(), require('./ruleofcode.config.json'), 'testing_557c4854')
// → 27 files
```

Its own entry point — the one the law calls — is right too:

```js
C.checkTestFilesExistence(process.cwd(), config, 'testing_557c4854')
// violations:
//   48 source files lack corresponding test files
//   Low test coverage ratio: 43.5% (27 tests for 62 source files, minimum 50%)
```

`43.5%` and `27 tests`. That is the correct reading, and the new
`describeCountedTests` suggestion lists them, which is a good addition.

## The audit reports the old numbers

Same repository, same version, same command we always run:

```
npx ruleofcode audit --mode=full
→ 62 source files lack corresponding test files
→ Low test coverage ratio: 1.6% (1 tests for 62 source files, minimum 50%)
```

`1.6%` and `1 test`. Note the source count agrees (62), so `projectRoot` is the
same in both paths — only the test side differs.

## What I ruled out

- **Version.** `npx ruleofcode --version` → 7.21.0.
  `require.resolve('ruleofcode/package.json')` → the top-level install. Exactly
  one file in the package contains the `Low test coverage ratio` template.
- **Caches.** `rm -rf .ruleofcode-cache node_modules/.cache`, then re-ran. Same.
- **lawId.** Tried `testing_557c4854`, `law-80`, `60`,
  `test-coverage-constitutional-standard` and `undefined` — 27 every time.
- **Config shape.** Tried our config, `{}`, ignores-only, and ours with
  `ignores` deleted — 27 every time. Our config has no `includes`, no
  `ignores.tests`, and `ignores.byRule` is empty.
- **The law's wiring.** `checkers/testing-laws/test-coverage-constitutional-standard.js`
  line 21 calls `TestFilesExistenceChecker.checkTestFilesExistence(context.projectRoot,
  context.config, context.lawId)` — the function I invoked.

That leaves `context.config` as the only variable I cannot inspect from outside:
the resolved config the runner builds, rather than the file on disk. Something in
it reduces test discovery to one file while leaving source discovery at 62.

`FileFilterUtils` reads several buckets we do not set — `ignores.tests`,
`ignores.build`, `ignores.design`, `includes.global`, `includes.byRule` — so a
default injected into any of those on the resolved config would do it, and would
be invisible to an adopter reading their own config.

## Repro

```bash
npm i ruleofcode@7.21.0
rm -rf .ruleofcode-cache node_modules/.cache
npx ruleofcode audit --mode=full | grep "test coverage ratio"
node -e "const {TestFilesExistenceChecker:C}=require('ruleofcode/dist/utils/testing/test-files-existence-checker.js');
         console.log(C.checkTestFilesExistence(process.cwd(), require('./ruleofcode.config.json'), 'testing_557c4854').violations)"
```

Two different answers from one version.

## Not urgent for us

The law is parked at `warning` with a note pointing here, and the note now says
the checker is right and the wiring is not. We are not blocked — I would rather
you have the precise shape than a quick patch.

**Comment by ruleofcodedev** [comment]
Re-tested against 7.21.0 (this was closed on 2026-08-18, before that release shipped). **The behaviour is unchanged**, so I am reporting rather than assuming the fix landed elsewhere.

## Still reported

```
13. Test Coverage Constitutional Standard [2 violations] (Score: 70/100)
   • 62 source files lack corresponding test files
   • Low test coverage ratio: 1.6% (1 tests for 62 source files, minimum 50%)
```

## The same run reads those files

`unit-test-quality-standards` in that identical audit names **18 distinct `.spec.ts` files** in its violations, one per file, and read more than that (only the failing ones are listed). So the tool walks them; one law counts them and the other counts one.

That is the same shape as #103 part 1 — two analyzers, one repository, opposite verdicts in a single run.

## Replicating the checker's own arithmetic

Its predicates and directory skips, copied verbatim from `test-files-existence-checker.js`, walked over this repository:

```
test files matching isTestFile   : 30
source files matching isSourceFile: 72
the law reports                  : 1 tests for 62 source files
```

The source side is right — 72 minus `shouldHaveTest` filtering gives 62. Only the test side is wrong, and it is wrong by 29.

## Our config cannot explain it

```jsonc
"includes": undefined              // no include scoping at all
"ignores": {
  "global": ["**/node_modules/**", "**/dist/**", "**/coverage/**", "**/.nx/**",
             "**/.angular/**", "**/.ruleofcode-cache/**", "**/tmp/**",
             "**/*vite*/**", "**/*chunk*/**"],
  "byRule": {}                     // empty
}
```

And reading the path it takes, nothing there should exclude them either. `findTestFiles` passes `includeTests: true`, and with that flag `getIgnorePatterns` does not add `SPEC_FILES`, `TEST_FILES`, `**/*-e2e/**`, `**/e2e/**` or `**/test/**` to the base patterns — it also strips those same patterns out of `ignores.global` before using it. `includes.global` is empty, so the include branch returns early. By reading, all 30 survive.

## Why it matters more than the number

The ratio cannot be moved by writing tests, which is what the law asks for. This repository has grown to 41 unit tests across 8 Jest suites and 209 Playwright checks since the issue was filed, and the reported figure has not moved off `1`.

Two smaller things noticed while reading:

- The message says `${testFiles.length} tests` while `testFiles` is a count of **files**. Even correct, "1 tests" would mean one file, not one test.
- `62 source files lack corresponding test files` is matched by basename after stripping `.spec`/`.test`, so a spec that lives beside its source still has to be found by the same walk — the two findings share the failing step.

Happy to run anything specific against this tree if it helps narrow it.

**Comment by ruleofcodedev** [comment]
Thank you for looking, and the cache defect you describe is real — I confirmed the key fields on our side before deleting it: `results, timestamp, filesHash, configHash, gitState, mode`, with no version anywhere. Putting the version in the key is right.

**It is not what produces this number, though.** I cleared it exactly as you asked and the figure did not move.

## What I ran

```console
$ mv .ruleofcode-cache <elsewhere>          # then: performance.cache = false
$ rm -rf .ruleofcode-cache
$ npm run audit
   • Low test coverage ratio: 1.6% (1 tests for 62 source files, minimum 50%)
$ ls .ruleofcode-cache
   (absent — cache: false is honoured, nothing recreated)
```

## Where our two measurements diverge

You are right that the checker finds all 30. So do I — **when it is handed our config file**:

```js
const { TestFilesExistenceChecker: C } = require('.../test-files-existence-checker.js');
C.findTestFiles(root, require('./ruleofcode.config.json'), undefined).length   // -> 30
```

The audit does not hand it that. It hands it the **loaded** config, and the loader injects a key our file does not contain:

```js
raw file has ignores.tests?  ->  false
LOADED ignores.tests         ->  ["**/*.spec.ts","**/*.spec.js","**/*.test.ts",
                                  "**/*.test.js","**/e2e/**","apps/*-e2e/**"]
LOADED includes              ->  {"global":[],"byRule":{}}
```

Same function, same tree, the only difference being that injected key:

```
config WITHOUT ignores.tests  ->  30 test files
config WITH    ignores.tests  ->   1 test file
survivors: [ 'apps/web/tsconfig.spec.json' ]
```

## The mechanism

`getIgnorePatterns` takes real care over `includeTests` and then undoes it two lines later:

```js
if (!includeTests) {
  basePatterns.push(PATTERNS.SPEC_FILES, PATTERNS.TEST_FILES, '**/cypress/**',
                    '**/*-e2e/**', '**/e2e/**', '**/test/**', '**/tests/**');
}
...
if (_config.ignores?.global?.length > 0) {
  if (includeTests) {
    // strips '**/*.spec.ts', '**/*.test.ts', '**/e2e/**' … out of the user's global ignores
    patterns.push(...filteredGlobalIgnores);
  } else { ... }
}
...
if (_config.ignores?.tests?.length > 0) {
  patterns.push(..._config.ignores.tests);      // <-- unconditional
}
```

`findTestFiles` asks with `includeTests: true`. The function keeps spec patterns out of the base list, scrubs them out of `ignores.global`, and then appends `ignores.tests` regardless — which is precisely that list of spec patterns, straight from `DEFAULT_CONFIG`.

So every real spec is ignored. The one file that survives is `apps/web/tsconfig.spec.json`: `isTestFile` matches on the substring `.spec.`, and the ignore patterns all end in `.ts`/`.js`, so a `.json` slips through. **The reported "1 tests" is one tsconfig.**

That also explains the second violation, `62 source files lack corresponding test files` — the pairing walks the same empty list, so no source can ever find its spec.

## Why our runs disagreed

Nothing to do with our tree. `ignores.tests` is in `DEFAULT_CONFIG` (`config/types.js:26`), with `apps/*-e2e/**` added in the Angular preset at line 144 — so any consumer who does not define `ignores.tests` gets it, and any run that bypasses the loader does not. Calling the checker directly, as I first did, gives 30 and looks fine.

## Suggested fix

One guard:

```js
if (!includeTests && _config.ignores?.tests?.length > 0) {
  patterns.push(..._config.ignores.tests);
}
```

Reproduction script, if it is useful — it needs only the tool and a config file:

```js
const { TestFilesExistenceChecker: C } = require('.../test-files-existence-checker.js');
const raw = require('./ruleofcode.config.json');
const withDefaults = structuredClone(raw);
withDefaults.ignores.tests = ['**/*.spec.ts','**/*.spec.js','**/*.test.ts','**/*.test.js','**/e2e/**'];
console.log(C.findTestFiles(process.cwd(), raw, undefined).length);          // 30
console.log(C.findTestFiles(process.cwd(), withDefaults, undefined).length); // 1
```

No apology needed on your side — you gave me a testable claim with a number in it, which is what let me find the actual path in one pass.

One unrelated thing the exchange did fix on our end: our config carried a comment saying `cache: false — a cached compliance result can report a stale PASSED`, while `performance.cache` was `true`. The comment described a decision that was not in force. It is now.
